include/envoy/http/filter.h

```diff
@@ -395,6 +395,13 @@ class StreamEncoderFilterCallbacks : public virtual StreamFilterCallbacks {
    */
   virtual void addEncodedData(Buffer::Instance& data, bool streaming_filter) PURE;
 
+  /**
+   * Provides a trailer map that can be used to modify the trailers for a response. Used when
+   * trailers need to be injected into a response and the upstream response did not contain any
+   * trailers (in which case encodeTrailers is not called).
+   */
+  virtual HeaderMap& addEncodedTrailers() PURE;
```

Comment:
unsure what the best signature was here - it seems like `HeaderMap` has a relatively rich API to allow calls to specify reference/copy/etc, so I wasn't sure how best to allow that without providing access to the `HeaderMap` itself. On the other hand, giving out an actual reference to the map might not be ideal as it opens up for more potential bugs (filter implementations keeping it around for too long). thoughts?

Reply:
How does this work if the target stream protocol is not able to support trailers?

Reply:
From what I understand the http/1.1 codec will just ignore the trailers: https://github.com/envoyproxy/envoy/blob/master/source/common/http/http1/codec_impl.cc#L148

Reply:
Perhaps we should just pass a HeaderMapPtr via move? This would just overwrite existing trailers? It would be uyp to the user to figure out if trailers already exist? Thoughts?

Reply:
> Perhaps we should just pass a HeaderMapPtr via move? This would just overwrite existing trailers? It would be uyp to the user to figure out if trailers already exist? Thoughts?

In that case you'll leaking HeaderMapImpl into filter codes and let filter create the HeaderMapPtr. Can we somehow (e.g. ASSERT) make sure this can be only called after or in `encodeData(data, end_stream=true)`? So the filter doesn't have to care about potentially overwriting existing (future) trailers.

Reply:
We will want to make sure for any protocol which doesn't support trailers we don't then lose the end_stream=true passing through the pipeline, but that's what tests are for :-)